### ISSUE TITLE
Elevated number of Timed out fetching a new connection from the pool

### ISSUE THREAD

**Issue by Sytten** [prisma/prisma#5977]
## Bug description

Related https://github.com/prisma/prisma/issues/5976 but warrant of its own issue.
There seems to be a regression since I updated clients from 2.14 to 2.17 in the connection pool.
I get a lot of timeout errors for a very similar traffic load:
```
Error: Timed out fetching a new connection from the pool. Please consider reducing the number of requests or increasing the `connection_limit` parameter (https://www.prisma.io/docs/concepts/components/prisma-client/connection-management#connection-pool). Current limit: 10.
    at cb (/app/node_modules/.prisma/client/runtime/index.js:78689:17)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async middleware (/app/node_modules/graphql-shield/dist/generator.js:30:24)
```

I highly suspect a bug in the code that returns the connections to the pool because I can see in my logs that it was the only executing query at the time of the timeout.

<img width="657" alt="Screen Shot 2021-03-04 at 1 52 11 PM" src="https://user-images.githubusercontent.com/2366731/110014497-ddcb1000-7cf0-11eb-8b9e-bce79e84fe1a.png">

## How to reproduce

Hard to reproduce.

## Expected behavior

Connections are properly returned to the pool.

## Version

| Name            | Version            |
|-----------------|--------------------|
| Node            | v14.15.4           | 
| OS              | debian-openssl-1.1.x|
| Prisma Client   | 2.17.0             |
| Query Engine    | query-engine 3c463ebd78b1d21d8fdacdd27899e280cf686223|
| Database        | undefined|

**Comment by Sytten** [comment]
Note that this project doesn't use subscription or anything really that can keep a request on the long term. And it shows that previous queries have all finished.

**Comment by janpio** [comment]
You could increase the pool timeout (e.g. `&pool_timeout=10`) or even disable it (`&pool_timeout=0`) to see if there are other errors hiding behind the pool timeout. Would be super interesting if this is the case.

**Comment by Sytten** [comment]
Hard sell on a production deployment deployment, will see what I can do.

**Comment by s-kris** [comment]
I migrated from 2.13.1 to 2.18.0 yesterday and since then, the connection limit errors occurring during my dev. It **never** happened in 2.13.1. **Never**. So, I temporarily downgrading. 

Earlier I was using serverless approach with nextjs but switched to custom server approach.  I instantiate prisma when server starts and pass down prisma client to all routes via req object.

**Comment by janpio** [comment]
Please open a separate issue @s-kris and describe your setup, code and situation a bit more. The error message you are seeing was only added recently, before it would have been an undefined error or crash that would not have given any indication. But we might also have changed some behavior by accident, so it's important to investigate.
(My first response will be the same as https://github.com/prisma/prisma/issues/5977#issuecomment-791402517 by the way - so if you can try that, this would be super helpful.)

**Comment by janpio** [comment]
@Sytten Increasing the pool timeout should not be a big problem, the existing default of `5` is already quite drastic and should really not be hit at any time.

**Comment by s-kris** [comment]
@janpio Sure, I'll open separate issue with more details and code. Thank you!

**Comment by Sytten** [comment]
Ok I digged a bit more into mobc/quaint and here is what I think should be part of the solution:
1. Expose `max_lifetime` on the Quaint builder https://github.com/prisma/quaint/blob/d72b591a31ab2af4ff8d7ed0fe91fc48bbc97deb/src/pooled.rs#L185
2. Set the default `max_lifetime` to 1h
3. Expose `max_lifetime` and `max_idle_lifetime` as query params for further tweaks by users if need be

The default `max_idle_lifetime` is already a 5m which is correct, but according to the response of my query on cloud run GCP can kill connection after 2minute, thus the need for a query parameter:
- Answer: https://issuetracker.google.com/issues/181655427#comment2
- Doc: https://cloud.google.com/appengine/docs/standard/go111/sockets

For my personal use case on other projects, I found that the lifetime of 1h on the connection helps a lot to reduce the bad connection issues.

I still think that there might be an underlining bug in the way the timeout is implemented. Maybe the request is not properly canceled (wild guess) and that means the connection is never returned to the pool. I say that because it seems the pool never recovered fully by itself when those timeout occur  (I had to force a restart of the service).

**Comment by Sytten** [comment]
After investigating mobc deeper, I found two bugs in the way the requests are handled and a PR is underway.
I still think the proposals above are relevant even with those fixes.

**Comment by matthewmueller** [comment]
@Sytten @s-kris can you help me understand the problem a bit better. Right now it's not super clear.

High query-per-second with a low connection_limit will cause timeouts. This is expected.

     Your query requests are waiting for a limited amount of connections to free up. Eventually they timeout. Increasing the connection_limit is a temporary solution, but the real solution is to upgrade your database.

I assume from this issue that this isn't your problem. That you don't have high query-per-second and you're still seeing timeouts. And did you try increasing the `connection_limit`?

**Comment by s-kris** [comment]
Hi @matthewmueller


No I didn't increase `connection_limit` but changing `pool_timeout` seems to get rid of the error. 

https://github.com/prisma/prisma/issues/6014#issuecomment-795589315

**Comment by Sytten** [comment]
Its on the way of being resolved, did a bunch of fixes already. I suggest you sync with @pimeys .

**Comment by janpio** [comment]
The eneral reason for increased amount of `Timed out fetching a new connection from the pool.` error messages should be gone since a release or two. This was due to a race between the connection timeout and the pool timeout which both defaulted to 5s.

The specific problem of @Sytten we are tackling as well - and will open a new issue with the then current error message if it is not actually resolved.

**Comment by Sytten** [comment]
There was also a missing FIFO in the connection pool requests that was solved in mobc 0.7.1.
